torchsde/_core/methods/milstein.py

```diff
@@ -12,30 +12,74 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from ...settings import SDE_TYPES, NOISE_TYPES, LEVY_AREA_APPROXIMATIONS
+import abc
+import math
+
+import torch
+
+from ...settings import SDE_TYPES, NOISE_TYPES, LEVY_AREA_APPROXIMATIONS, METHOD_OPTIONS
 
 from .. import base_solver
 
 
-class Milstein(base_solver.BaseSDESolver):
+class BaseMilstein(base_solver.BaseSDESolver, metaclass=abc.ABCMeta):
     strong_order = 1.0
     weak_order = 1.0
-    sde_type = SDE_TYPES.ito
     noise_types = (NOISE_TYPES.additive, NOISE_TYPES.diagonal, NOISE_TYPES.scalar)
     levy_area_approximations = LEVY_AREA_APPROXIMATIONS.all()
 
+    def __init__(self, options, **kwargs):
+        if METHOD_OPTIONS.grad_free not in options:
+            options[METHOD_OPTIONS.grad_free] = False
+        super(BaseMilstein, self).__init__(options=options, **kwargs)
+
+    @abc.abstractmethod
+    def v_term(self, I_k, dt):
+        raise NotImplementedError
+
     def step(self, t0, y0, dt):
         assert dt > 0, 'Underflow in dt {}'.format(dt)
 
-        I_k = self.bm(t0, t0 + dt)
-        v = [delta_bm_ ** 2. - dt for delta_bm_ in I_k]
+        t1 = t0 + dt
+
+        I_k = self.bm(t0, t1)
+        v = self.v_term(I_k, dt)
 
         f_eval = self.sde.f(t0, y0)
-        g_prod_eval = self.sde.g_prod(t0, y0, I_k)
-        gdg_prod_eval = self.sde.gdg_prod(t0, y0, v)
+        g_eval = self.sde.g(t0, y0)
+        g_prod_eval = self.sde.prod(g_eval, I_k)
```

Comment:
`prod` will break the adjoint code. I don't think we should optimize this for now, and I'd rather just use `g_prod` at the cost of a few extra function evaluations. Feel free to add this to #23.

Reply:
Sure! Will reverse it.

Reply:
Thanks! We can brainstorm about ways to speed things up after this PR gets merged. On a high-level, the reason we made `g_prod` in the first place is that it's a vjp for the adjoint, and vjps are faster than directly computing the Jacobian. Evaluating g directly for the adjoint would mean computing the whole Jacobian, which scales poorly w.r.t. the dimension.

Reply:
Sure thing!
For now I just pushed changes.